Update Readme with latest changes (#324)
submariner-io/submariner: README.md

The basic architecture diagram for Submariner is as follows:

![submariner architecture](docs/img/architecture.jpg)

Submariner consists of a few components that work and operate off of Kubernetes Custom Resource Definitions (CRDs). The Submariner CRDs are `clusters.submariner.io` and `endpoints.submariner.io`. 

The two primary Submariner components within connected clusters are:

- submariner-gateway (DaemonSet)
- submariner-route-agent (DaemonSet)

> The `submariner-gateway` pods are run on the gateway labelled nodes, and will perform leader election between them to elect an active IPsec endpoint. `submariner-route-agent` as a component is run on every node, and is aware of the current leader. While running on the gateway node that is the leader, it will create a VxLAN VTEP interface to which all the worker nodes are connected, whereas on the non-leader nodes, it will setup a VXLAN tunnel with remote VTEP as the active gateway node, configures routes, programs necessary iptable rules and enables full connectivity to the remote clusters. The MTU of Submariner VxLAN tunnel is configured based on the MTU of the default interface on the host minus VxLAN overhead.

Reviewer: ```suggestion
The `submariner-gateway` pods are run on the gateway-labeled nodes and perform a leader election between them to elect an active IPsec endpoint. The `submariner-route-agent` component runs on every node and is aware of the current active gateway. When running on the active gateway node, the `submariner-route-agent` creates a VxLAN VTEP interface to which the `submariner-route-agent` instances running on the other worker nodes connect by establishing a VXLAN tunnel with the VTEP of the active gateway node. The `submariner-route-agent` also configures routes and programs the necessary iptable rules to enable full connectivity to the remote clusters. The MTU of Submariner VxLAN tunnel is configured based on the MTU of the default interface on the host minus VxLAN overhead.
```

Author: Done

Upon startup, the `submariner-gateway` pod that is elected leader will perform a reconciliation process that ensures it is the sole endpoint for this cluster. This is part of the reason why it is important to have unique cluster IDs between clusters, as two clusters with the same ID will reconcile each other out of existence.

> Upon failure, another `submariner-gateway` pod (on one of the other gateway hosts) will gain leadership and perform reconciliation to ensure it is the active leader. When done, the remote clusters will reconcile the IPsec endpoint to the new endpoint, and connection will be re-established. In the interim, the `submariner-route-agent` pods will update the route tables on each node to point towards the new endpoint host.

Reviewer: ```suggestion
Upon failure, another `submariner-gateway` pod (on one of the other gateway nodes) will gain leadership and perform reconciliation to ensure it is the active leader. When done, the remote clusters will reconcile the IPsec endpoint to the new endpoint, and connection will be re-established. In addition, the `submariner-route-agent` pods will update the route tables on each node to point to the new active gateway node.
```

Author: Done

Submariner uses a central broker to facilitate the exchange of information and sync CRD's between clusters. The `datastoresyncer` runs as a controller within the leader-elected `submariner` pod, and is responsible for performing a two-way synchronization between the datastore and local cluster of Submariner CRDs. The `datastoresyncer` will only push CRD data to the central broker for the local cluster (based on cluster ID), and will sync all data from the broker the local cluster when the data does not match the local cluster (to prevent circular loops)

## submariner-gateway

submariner-gateway (compiled as the binary `submariner-engine`) has a few controllers built into it that establish state. It is responsible for running/interfacing with Charon to establish IPsec tunnels, as well as updating local cluster information into the central broker to share information between clusters.

> submariner-gateway runs as a DaemonSet on the gateway labelled nodes and utilizes leader election to establish an active gateway node, which is used to facilitate IPsec tunnel connections to remote clusters.

Reviewer: ```suggestion
submariner-gateway runs as a DaemonSet on the gateway-labeled nodes and utilizes leader election to establish an active gateway node, which is used to facilitate IPsec tunnel connections to remote clusters.
```

Author: Done

When the source pod is on a worker node that is not the elected gateway node, the traffic destined for the remote cluster will transit through the submariner VxLAN tunnel (`vx-submariner`) to the local cluster gateway node. On the gateway node, traffic is encapsulated in an IPSEC tunnel and forwarded to the remote cluster. Once the traffic reaches the destination gateway node, it is routed in one of two ways, depending on the destination CIDR. If the destination CIDR is a pod network, the traffic is routed via CNI-programmed network. If the destination CIDR is a service network, then traffic is routed through the facility configured via kube-proxy on the destination gateway node.

# Prerequisites

Submariner has a few requirements in order to get started:

- At least 2 Kubernetes clusters, one of which is designated to serve as the central broker that is accessible by all of your connected clusters; this can be one of your connected clusters, but comes with the limitation that the cluster is required to be up in order to facilitate interconnectivity/negotiation

> - Different cluster/service CIDR's (as well as different kubernetes DNS suffixes) between clusters. This is to prevent traffic selector/policy/routing conflicts.

Reviewer 2: I would also add, - Worker node IPs on all cluster must be outside of the cluster/service CIDR ranges.

Author: Sure, I will add it. Thanks for reviewing the PR @mangelajo
